[FEAT] 업로드 세션 시작(StartUpload) 기능 구현

juyeongMoon888 opened the pull request:
## 📝 작업 내용
- 업로드 세션 생성을 위한 `POST /api/uploads` API 구현
- 업로드 세션 시작을 담당하는 `StartUploadUseCase` 추가
- 업로드 세션 도메인 모델(`UploadSession`) 및 상태(`UploadSessionStatus`) 정의
- 업로드 세션 영속성 계층 구성
  - `UploadSessionRepository`
  - `UploadSessionJpaRepository`
  - `UploadSessionRepositoryImpl`
- 업로드 관련 설정을 외부화하기 위한 `UploadProperties` 설정 클래스 추가
- 공용 `application.yml`에 업로드 세션 관련 설정 추가
  - 임시 업로드 디렉터리(`upload.temp.base-dir`)
  - 업로드 세션 TTL(`upload.session.ttl-hours`)
  - 청크 최대 크기(`upload.chunk.max-size-bytes`)
- JPA Auditing 및 `UploadProperties` 설정 바인딩 활성화
- 업로드 세션 시작 API 검증을 위한 HTTP 테스트 스크립트 추가

---

## 💭 주의 사항
- 본 PR은 **업로드 세션 생성(Start 단계)** 까지만 포함합니다.
- 실제 파일 바이트 전송(청크 업로드) 로직은 포함되어 있지 않습니다.
- `tempFilePath`는 서버 내부 구현 정보이므로 API 응답에 포함하지 않았습니다.
- 업로드 세션의 초기 상태는 생성 시점에 `UPLOADING`으로 고정되며,
  이후 상태 변경은 도메인 메서드를 통해 처리할 예정입니다.
- 업로드 세션 만료 여부는 서버 기준으로 판단하며,
  재개 처리를 위한 조회 API는 후속 작업에서 추가 예정입니다.
- 🔥예외처리 응답포맷은 추후 구현예정입니다. 

---

## 💡 리뷰 포인트
- StartUploadUseCase에서의 입력 검증 책임 분리 방식이 적절한지
- 업로드 세션 생성 시 도메인 팩토리(`start`) 방식이 적절한지
- `UploadProperties` 설정 구조가 향후 확장(예: S3 Multipart 업로드)에 적합한지
- StartUpload API 응답에 포함된 필드  
  (`declaredTotalSize`, `nextOffset`, `maxChunkSize`)의 적절성
- 공용 설정(`application.yml`)에 추가된 업로드 관련 설정 네이밍 및 범위

## ✅ 테스트
- [x] 로컬 빌드/실행
- [ ] 단위 테스트 추가/통과
- [x] API 수동 테스트
<img width="873" height="338" alt="스크린샷 2025-12-28 오후 9 59 23" src="https://github.com/user-attachments/assets/5d8248b4-f00d-4383-b7ae-fde220f89419" />

juyeongMoon888 committed:
feat: 업로드 세션 시작 API 컨트롤러 및 요청/응답 DTO 구현

juyeongMoon888 committed:
feat: 업로드 세션 설정 및 Repository 구현

juyeongMoon888 committed:
feat: 업로드 세션 도메인 모델 및 상태 정의

juyeongMoon888 committed:
feat: 업로드 세션 시작 유스케이스 구현

juyeongMoon888 committed:
test: dj업로드 세션 시작 API HTTP 요청 예시 추가

juyeongMoon888 committed:
chore: JPA Auditing과 ConfigurationProperties 활성화

sanseongKo commented on src/main/java/com/sansung_gorogoro/file_server/presentation/controller/UploadController.java:
```diff
@@ -0,0 +1,28 @@
+    private final StartUploadUseCase startUploadUseCase;
+
+    @PostMapping("/uploads")
+    public ResponseEntity<?> create (@Valid @RequestBody StartUploadRequest req,
```
제네릭 안에 어떤 타입이 명확히 들어간다면 그렇게 바꿔주세요!

sanseongKo commented on src/main/java/com/sansung_gorogoro/file_server/presentation/request/StartUploadRequest.java:
```diff
@@ -0,0 +1,20 @@
+import jakarta.validation.constraints.Size;
+import lombok.Builder;
+
+@Builder
```
빌더를 따로 사용하신 이유가 잇을까요?

sanseongKo commented on src/main/java/com/sansung_gorogoro/file_server/presentation/request/StartUploadRequest.java:
```diff
@@ -0,0 +1,20 @@
+@Builder
+public record StartUploadRequest(
+        @NotBlank(message = "originalFileName은 필수입니다.")
+        @Size(max = 255)
```
이 부분에도 메세지가 따로 있으면 좋을 것같아요!

sanseongKo commented on src/main/java/com/sansung_gorogoro/file_server/presentation/response/StartUploadResponse.java:
```diff
@@ -0,0 +1,23 @@
+import java.time.LocalDateTime;
+
+@Builder
+public record StartUploadResponse(
```
빌더가 필요한 이유가 있을까요?

juyeongMoon888 replied:
StartUploadResponse는 DTO 특성상 필드가 확장될 가능성이 높고,
생성자 기반`(new StartUploadResponse(...))` 방식은 파라미터 순서 오류를 유발하기 쉽습니다.

이에 가독성과 안전성을 높이기 위해 `Builder` 패턴을 적용했으며,
필드 추가 시에도 호출부 변경을 최소화할 수 있도록 설계했습니다.😀

sanseongKo commented on src/main/java/com/sansung_gorogoro/file_server/infrastructure/config/UploadProperties.java:
```diff
@@ -0,0 +1,14 @@
+
+@ConfigurationProperties(prefix = "upload")
+public record UploadProperties(
+        Temp temp,
```
네이밍이 조금 명확하지 않은 것 같아요! 또한, 객체로 받으시는 이유가 있으신가요?

juyeongMoon888 replied:
**네이밍:** 
말씀 주신대로 UploadProperties가 범위가 넓어 모호할 수 있어
의도가 드러나게 수정하겠습니다. (temp, session, chunk 이름들도 설정 타입과 단위를 넣어 명확하게 수정하겠습니다.)

**객체로 받은 이유에 대해서는:** 
초기 설계 단계에서 여러 구현 방식을 참고하는 과정에서
중첩 객체 구조를 우선 적용했으나, 
해당 구조에 대한 명확한 필요성과 책임을 충분히 정의하지 못한 상태였습니다.

현재 단계에서는 설정 수와 복잡도를 고려해
중첩 객체 대신 **필드 구조**가 더 명확하다고 판단하여 수정하려고 합니다. 

감사합니다!

sanseongKo commented on src/main/java/com/sansung_gorogoro/file_server/domain/UploadSession.java:
```diff
@@ -0,0 +1,70 @@
+package com.sansung_gorogoro.file_server.domain;
+
+import jakarta.persistence.*;
```
와일드 카드는 제거해주세요!

sanseongKo commented on src/main/java/com/sansung_gorogoro/file_server/domain/UploadSession.java:
```diff
@@ -0,0 +1,70 @@
+import java.time.LocalDateTime;
+
+@Entity
+@Getter
```
이건 살짝 취향 차이이긴한데 게터가 전부 필요한지도 한번 생각해보면 좋을거같아요

juyeongMoon888 replied:
조회/재개/확정 API 설계 확정 후, 실제 사용되는 필드 기준으로 getter 노출 범위를 부분 적용할 예정입니다!

sanseongKo commented on src/main/java/com/sansung_gorogoro/file_server/domain/UploadSession.java:
```diff
@@ -0,0 +1,70 @@
+public class UploadSession {
+    @Id
+    @Column(name = "upload_id", length = 36, nullable = false, updatable = false)
+    private String uploadId; // UUID string
```
mysql은 uuid 타입을 제공하는데 스트링으로 받으시는 이유가 있을까요?

juyeongMoon888 replied:
DB/운영 환경에 따라 UUID 저장 방식(BINARY, CHAR 등) 차이로 인한 호환성 이슈를 피하기 위해 문자열로 관리했었습니다.
다만 현재 프로젝트 범위에서는 UUID의 전역 유니크 특성이 필요하지 않아, 이번 PR에서 식별자 전략을 Long(Id) 기반으로 변경하겠습니다. 
조언 감사합니다😀

sanseongKo commented on src/main/java/com/sansung_gorogoro/file_server/domain/UploadSession.java:
```diff
@@ -0,0 +1,70 @@
+                                      String originalFileName,
+                                      String tempFilePath,
+                                      LocalDateTime expiresAt) {
+        return new UploadSession(uploadId, ownerUserId, declaredTotalSize, originalFileName, tempFilePath, expiresAt);
```
해당 도메인에는 별다른 불변식은 존재하지 않는걸까요?

sanseongKo commented on src/main/java/com/sansung_gorogoro/file_server/application/result/StartUploadResult.java:
```diff
@@ -0,0 +1,22 @@
+import com.sansung_gorogoro.file_server.domain.UploadSession;
+import lombok.Builder;
+
+@Builder
```
이쪽도!

juyeongMoon888 replied:
StartUploadResult 역시 유스케이스의 반환 DTO로서 필드 확장 가능성이 있고,
생성자 기반 생성은 파라미터 순서 오류 유발 문제가 있습니다.
Builder를 적용해 호출부 가독성과 안전성을 높였고, 필드 추가 시 변경 범위를 최소화했습니다.😀

sanseongKo commented on src/main/java/com/sansung_gorogoro/file_server/application/usecase/StartUploadUseCase.java:
```diff
@@ -0,0 +1,55 @@
+        try {
+            Files.createDirectories(baseDir);
+        } catch (IOException e) {
+            throw new IllegalStateException("임시 업로드 디렉터리를 생성할 수 없습니다: " + baseDir, e);
```
예외는 이넘타입으로 따로 만들어 두시는건 어떨까요?

sanseongKo replied:
다른 프로젝트를 한번 확인해주세요!

juyeongMoon888 replied:
해당 PR에서는
공통 응답 포맷 및 예외 처리 적용은 포함하지 않았습니다.
응답 포맷 관련 내용은 별도의 이슈(응답 포맷 이슈)에서 공통 응답/에러 포맷을 정의한 뒤,
해당 이슈에서 예외 및 응답 포맷을 일괄 적용할 예정입니다.

sanseongKo commented on src/main/java/com/sansung_gorogoro/file_server/application/usecase/StartUploadUseCase.java:
```diff
@@ -0,0 +1,55 @@
+
+        Path tempFilePath = baseDir.resolve(TEMP_PREFIX + uploadId + TEMP_EXT);
+
+        LocalDateTime expiresAt = LocalDateTime.now().plusHours(props.session().ttlHours());
```
expireAt을 서비스레이어에서 잡아주는 의도가 따로 존재하시는거죠?!

juyeongMoon888 replied:
전체 흐름을 빠르게 잡는 데 집중하다 보니,만료 시간 계산을 서비스 레이어에서 직접 처리했습니다.
이후 해당 로직이 운영 정책 성격임이 명확해져, 책임을 분리하는 방향으로 변경했습니다!

sanseongKo commented on http/files/upload-session-api.http:
```diff
@@ -0,0 +1,11 @@
+{
+  "originalFileName": "215926exam1234561111_small.mp4",
+  "declaredTotalSize": 12345678
+}
```
해당 파일은 안올라가게 해주세요! 코드에 필요하지 않은 파일인 것 같아요

sanseongKo commented on src/main/resources/application.yml:
```diff
@@ -3,4 +3,10 @@ spring:
+  session:
+    ttl-hours: 24
+  chunk:
+    max-size-bytes: 5242880
```
yml파일이 올라간다면, 변수로 저장해주세요 불변값을 넣어두면 나중에 실제 돌릴때 값을 변경할 수 없어요~

juyeongMoon888 replied:
multipart 및 upload 관련 설정을 기본값이 있는 외부 변수(${}) 형태로 분리해
환경별로 유연하게 조정 가능하도록 수정하겠습니다!

말씀 주신 설정 관련 사항은 환경설정 PR에서 반영했습니다!

sanseongKo commented on src/main/java/com/sansung_gorogoro/file_server/FileServerApplication.java:
```diff
@@ -1,13 +1,17 @@
+import org.springframework.data.jpa.repository.config.EnableJpaAuditing;
 
+@EnableJpaAuditing
+@EnableConfigurationProperties(UploadProperties.class)
```
스프링어플리케이션에 해당 설정을 둔 이유가 있으신가요?!

juyeongMoon888 replied:
초기 구현 단계에서는 빠른 기능 검증에 집중하면서 애플리케이션 진입점에 설정을 함께 두었습니다.
다만 해당 설정들은 애플리케이션 부트스트랩 역할과는 분리하는 것이 책임 측면에서 더 명확하다고 판단되어, 말씀 주신 것처럼 config 전용 클래스로 분리하는 방향이 더 적절해 보입니다.!
피드백 감사합니다. 해당 구조로 정리하겠습니다.😀

sanseongKo reviewed (changes requested):
고생하셨습니다 👍 코멘트 확인해주세요~
틀렸다고 지적한 부분이 아니라 의도가 궁금했고 의도가 있으시다면 이해할 수 있습니다!
무조건적인 변경하겠습니다는 지양해주세요

sanseongKo commented:
그리고 커밋 의미에 맞지 않는 코드나 파일이 존재하는 것 같아요 다음엔 커밋을 보는 사람 입장에서 적어주시면 더욱 좋은 협업방식이 될 것 같습니다 :)

juyeongMoon888 committed:
[File] StartUploadRequest DTO 검증 메시지 보강 및 빌더 제거

juyeongMoon888 committed:
[File] upload 정책 설정 구성 및 검증 추가

juyeongMoon888 committed:
[File] 로컬 전용 HTTP 클라이언트 파일 저장소 제외 처리

juyeongMoon888 committed:
[File] 업로드 세션 생성 시 만료/경로 생성 책임 분리

juyeongMoon888 committed:
[File] style: UploadSession import 정리 및 컨트롤러 파라미터 컨벤션 정리

juyeongMoon888 committed:
[File] UploadSession 소유자 식별자 ownerUserId 타입을 계약에 맞춰 Long으로 변경

식별자 타입 정리 과정에서 임시 파일 경로 생성 시 잘못 전달되던 파라미터(ownerUserId)를 uploadId로 수정했습니다.

juyeongMoon888 committed:
[File] UploadSession 식별자에 IDENTITY 적용 및 임시 파일 경로 계산 방식 변경

- upload_session PK를 Long IDENTITY로 변경
- tempFilePath는 id 기반으로 재계산 가능한 파생 값이라 컬럼 저장하지 않고 경로 계산 방식으로 전환
- 이에 따른 저장/조회 흐름 정리

juyeongMoon888 committed:
[File] UploadSession 생성 시점 불변식 명시

sanseongKo reviewed (approved):
고생하셨습니다 👍
